lsp/CONTRIBUTING.md

```diff
@@ -0,0 +1,227 @@
+```bash
+pkg --compress GZip .
+```
+
+## Running + Debugging
+
+> **NOTE**: Ensure your VSCode workspace is `aws-toolkit-common/lsp` or else certain functionality may not work.
+
+### With Minimal VSCode Client
+
+This repo contains a minimal vscode client that can be used to easily run and
+debug changes to this language server.
+
+1. In the `Run & Debug` menu, run `"Launch as VSCode Extension + Debugging"`
+2. Set breakpoints in `src` where needed.
+
+### With VSCode Toolkit Extension
+
+The VSCode Toolkit Extension can start the AWS Documents Language Server itself.
+This will explain how to setup the extension to run with the language server
+and be able to debug it all.
+
+> **NOTE**: The cloned git repos must be adjacent to each other in the filesystem since commands are currently using relative paths.
+
+1. Clone the [`aws-toolkit-vscode`](https://github.com/aws/aws-toolkit-vscode) repo:
+    ```
+    git clone git@github.com:aws/aws-toolkit-vscode.git
+
+    cd aws-toolkit-vscode && \
+
+    git fetch origin jpinkney-aws/lsp && git checkout jpinkney-aws/lsp && \
+
+    cd ..
+    ```
+
+2. Clone the [`aws-toolkit-common`](https://github.com/aws/aws-toolkit-common) repo:
+    ```console
+    git clone git@github.com:aws/aws-toolkit-common.git && \
+
+    cd aws-toolkit-common && \
+
+    git fetch origin test-typescript-lsp && git checkout test-typescript-lsp && \
+
+    cd ..
+    ```
+
+3. Open each project in their own VSCode window.
+
+4. In `aws-toolkit-common` run:
+    ```console
+    npm run watch
+    ```
+
+5. In `aws-toolkit-vscode` start the extesion in `Run & Debug` using the `"Extension"` launch config.
+A new window will open.
+
+6. In the new window enable the language server in the VSCode settings under `aws.experiments` and check `lsp`. This will start the language server using the output you generated in step `4`.
+
+7. In the `aws-toolkit-common` VSCode window connect the debugger in `Run & Debug` using the `"Attach to AWS Documents Language Server"` launch config. Set breakpoints where needed.
+
+
+## Testing
+### Running Tests
+**In VSCode**
+
+In the `Run & Debug` menu are options to separately
+run unit or integration tests. Additionally, you
+can test the file that is currently open.
+
+**On CLI**
+
+How to test using the CLI.
+
+Unit Testing:
+```bash
+npm test
+```
+
+Integration Testing:
+```bash
+npm run testInt
+```
+---
+### Writing Tests
+
+* The modules used in testing are:
+    * `mocha`: Testing framework
+    * `chai`: For assertions
+    * `sinon`: stub/mock/spy
+
+* The design of the source code is written with [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection)
+in mind.
+    * An object or function receives other objects or functions
+    for functionality that it depends on, instead of all functionality
+    existing statically in one place.
+    * This simplifies testing and influences how tests are to be written.
+
+#### How To Use `sinon`
```

Comment:
I commend the effort, personally I would've just linked to `sinon` docs and/or examples in the codebase